feat: Register docs (#22722)
newrelic/docs-website: src/content/docs/browser/new-relic-browser/browser-pro-features/micro-frontend-monitoring.mdx

---
title: Monitor micro frontends with Browser monitoring
tags:
  - Browser
  - Browser monitoring
  - Micro frontends
metaDescription: "Monitor the performance and user experience of individual micro frontends within your web application using New Relic Browser monitoring."
freshnessValidatedDate: never
---

> Micro frontend (MFE) monitoring lets you track the performance and user experience of individual micro frontends within your web application. Instead of seeing all data aggregated at the page level, you can now isolate telemetry data for each micro frontend component into distinct entities in New Relic, giving you precise visibility into how each part of your application performs.

Reviewer: I'm saying this as well, because a very common way (and a Micro Frontend recommended practice) is to split a web application up is that micro frontends become pages or a set of pages that belong together. So mentioning: "Instead of page level aggregation" might be confusing...

## What are micro frontends? [#what-are-mfes]

Micro frontends are an architectural pattern where large web applications are broken down into smaller, independently deployable components. Each micro frontend can be:

- Owned and maintained by separate team(s)
- Developed using independent technologies or frameworks
- Deployed independently of other components
- Representative of a specific business domain or feature

> Common micro frontend patterns include:

Reviewer: I would maybe remove the list below. I'm not sure if this brings any value to the reader, unless they link to a deeper explanation somewhere in the docs or externally...

Micro frontend monitoring solves these challenges by providing:

- **Scoped timing data** for each micro frontend
- **Independent error and event tracking** with proper attribution
- **Team and Respository assignment** to solve ownership-based questions
- **Relationship mapping** between micro frontends and the container application
- **Granular control of custom data** at the component level

## How New Relic's micro frontend monitoring works [#how-it-works]

New Relic's approach uses a **central agent architecture** where:

1. **One browser agent** runs in the container application, which continues to capture page-wide data. This agent would typically be used by the container-level or platform teams, interested in monitoring holistic application performance.
2. **Each micro frontend registers** with the central agent using the `register` API
3. **Data is scoped and harvested** separately for each registered micro frontend when using the scoped APIs.
4. **New MicroFrontend entities automatically appear** in New Relic when utilized properly.
5. **Relationships are automatically created** between the container and micro frontends.

> 6. **Entity-level attibution is supported** to facilitate assignment of teams, repositories and tags.

Reviewer: ```suggestion
6. **Entity-level attribution is supported** to facilitate assignment of teams, repositories and tags.
```

- [`interaction()`](/docs/browser/new-relic-browser/browser-apis/interaction) - Access the SPA interaction API for custom timing
- [`log()`](/docs/browser/new-relic-browser/browser-apis/log) - Capture log messages scoped to your micro frontend
- [`measure()`](/docs/browser/new-relic-browser/browser-apis/measure) - Create custom performance measurements
- [`noticeError()`](/docs/browser/new-relic-browser/browser-apis/noticeerror) - Manually capture errors with micro frontend attribution
- [`register()`](/docs/browser/new-relic-browser/browser-apis/register) - Register an entity for scoped data observability
- [`recordCustomEvent()`](/docs/browser/new-relic-browser/browser-apis/recordcustomevent) - Record custom events for analytics
- [`setApplicationVersion()`](/docs/browser/new-relic-browser/browser-apis/setapplicationversion) - Track the specific version of your micro frontend
- [`setCustomAttribute()`](/docs/browser/new-relic-browser/browser-apis/setcustomattribute) - Add custom attributes that apply to all data from this micro frontend
- [`setUserId()`](/docs/browser/new-relic-browser/browser-apis/setuserid) - Associate a user ID with this micro frontend's data

> Each of these methods works identically to the standard browser agent APIs, but automatically scopes all collected data to your micro frontend entity.

Author: ```suggestion
### How registration works.
All micro frontend activity is facilitated by using the `register` API on the container browser agent. Please refer to the [register API](/docs/browser/new-relic-browser/browser-apis/register) documentation for more information.

### Key Requirements

> 1. IDs supplied should be generated and managed according to your organization's internal processes and conventions. It serves as a permanent, stable reference to uniquely identify this specific documentation entity across your system. Ensure this value remains consistent and is not duplicated across other documents or entities.

Reviewer: ```suggestion
1. IDs supplied should be generated and managed according to your organization's internal processes and conventions. It serves as a permanent, stable reference to uniquely identify this specific entity across your system. Ensure this value remains consistent and is not duplicated across other entities.
```

Not sure what is ment with "documentation" and "document" here?

  id: '550e8400-e29b-41d4-a716-446655440000',
  name: 'Catalog Viewer',
  tags: ['Documentation', 'Customer Facing']
});

// Set module-specific attributes, which will append attributes to all data collected about your micro frontend
catalogAgent.setCustomAttribute('catalogVersion', '2.1.0');
catalogAgent.setCustomAttribute('displayMode', 'grid');

// Set user context for this micro frontend
catalogAgent.setUserId('user-abc-123');

// Track the version of this specific micro frontend
catalogAgent.setApplicationVersion('catalog-mfe-v2.1.0');

// Track custom page actions
catalogAgent.addPageAction('catalogItemViewed', { 
  itemId: '12345',
  category: 'electronics' 
});

// Capture errors specific to this micro frontend
try {
  // Component logic that might fail
  loadCatalogItems();
} catch (error) {
  catalogAgent.noticeError(error, { 
    context: 'catalog-load', // custom attributes for more information
    failedOperation: 'fetchItems' // custom attributes for more information
  });
}

// Track custom logs specific to this module
catalogAgent.log('Catalog items loaded successfully', { 
  level: 'debug', customAttributes: {
    itemCount: 150,
    loadTime: 245 
  }

> });

Reviewer: ```suggestion
catalogAgent.log('Catalog items loaded successfully', { 
    level: 'debug', 
    customAttributes: {
      itemCount: 150,
      loadTime: 245 
    }
  }
);
```

// Record custom events for analytics
catalogAgent.recordCustomEvent('CatalogInteraction', {
  action: 'filter-applied',
  filterType: 'price-range',
  minPrice: 10,
  maxPrice: 100
});

// Measure custom performance metrics
catalogAgent.measure('catalog-rerender', {
  start: 1234,
  end: 5678
});

// Cleanup when the micro frontend unmounts and is no longer needed
catalogAgent.deregister();
```

> ## Generic component monitoring [#generic-components]

Reviewer: While technically possible... do you think we should document this?

We might run the risk that the Micro Frontend solution need to monitor everything, since everything you can think of can be scoped some way or another. I'm worried that this will make it really difficult to make a strong product solution. For example, you can "scope" also the process of prefetching data or the user click rage behaviour, but making that a separate entity seems a bit weird to me.

Like I said before, customers can do strange things with our agents and they most likely will. But I always thought it would be a good idea to have a strong thought leadership opinion so we can optimise our product solution around that.

Author: This was asked for already from customers and by our product leaders already which is why its included in here.  It may come to be removed before becoming public

- **Shared components** across your pages
- **Third-party widgets** embedded in your application
- **Feature flags** and A/B test variants
- **User journey segments** like onboarding flows
- **Dynamic content areas** that load different experiences

```js
// Monitor a third-party chat widget
const chatWidget = window.newrelic.register({
  id: '7f9e4a1b-c3d2-4e8f-9b5a-1d6c8e2f4a3b',
  name: 'Customer Support Chat',
  tags: ['Component', 'Customer Facing']
});

chatWidget.addPageAction('chatStarted');
chatWidget.setCustomAttribute('widgetVersion', '2.1.4');

// Monitor an onboarding flow
const onboardingAgent = window.newrelic.register({
  id: 'a1b2c3d4-e5f6-7890-abcd-ef1234567890',
  name: 'New User Onboarding',
  tags: ['Component', 'Internal']
});

onboardingAgent.addPageAction('onboardingStepCompleted', { step: 3 });
```

## Data relationships and entity synthesis [#relationships]

When you register a micro frontend, New Relic automatically:

1. **Creates a browser entity** for the micro frontend if it doesn't exist
2. **Establishes relationships** between the container and micro frontend entities

> 3. **Tracks various timings** describing the import and execution times of your micro frontend scripts

Author: ```suggestion
3. **Tracks various timings** describing the lifecycle timings of your micro frontend scripts
```

```js
const mfe1 = window.newrelic.register({
  id: 'a1b2c3d4-e5f6-7890-abcd-ef1234567890',
  name: 'Parent MFE'
})

// register another MFE as a child of mfe1
const mfe2 = mfe1.register({
  id: '7f9e4a1b-c3d2-4e8f-9b5a-1d6c8e2f4a3b',
  name: 'Child MFE'
})
```

These entities would then be related in New Relic like: `Browser Agent -- contains --> Parent MFE -- contains --> Child MFE` 

### View your related entities
The relationships appear in your service maps and entity explorer, showing how your micro frontends connect to your broader application architecture.

## Configuration at the container
For detailed implementation steps, see our [micro frontend configuration guide](/docs/browser/new-relic-browser/configuration/configure-micro-frontend-monitoring).

### Enablement
The container browser agent acts as the gatekeeper for the page and controls what any registered children can report. As such, it can be configured to block incoming registration data:

To configure this, edit the application settings in New Relic for your container agent app.  To do so with NPM, ensure that the container agent is supplied with:
```js
{
  init: {
    // ... other configs
    api: {

>       allow_registered_children: true|false

Author: ```suggestion
      allow_registered_children: true|false // defaults to `false`
```

    }
  }
}
```

## Data duplication

The container browser agent can be configured to duplicate data reported by registered micro frontends to its own app, thereby ensuring that _all_ data can be observed from the top-level application if desired.

> This can be further broken down, as the agent will automatically add `parent.id` and `child.id` attributes to the relevant data reported by each, enabling easy cross-linking as needed.

Reviewer: Will this parent and child ID only be added if duplicated data is enabled or is it always present?

To configure this, edit the application settings in New Relic for your container agent app.  To do so with NPM, ensure that the container agent is supplied with:
```js
{
  init: {
    // ... other configs
    api: {
      duplicate_registered_data: true|false
    }
  }
}
```

## Performance timing and measurements [#performance-timing]

Micro frontend monitoring can capture the same performance data as standard browser monitoring, but scoped to each component:

- **Performance timings** (when applicable to the micro frontend's initialization) - reported as `MicroFrontEndTiming` events

> - **AJAX requests** made from within the micro frontend context  

Reviewer: How are AJAX request being captured so they are scoped to the MFE? I believe this is still missing, right?

Author: yeah this needs to be omitted for now, ty

Author: ```suggestion
```

> - **JavaScript errors** that occur in micro frontend code

Reviewer: This might be misleading, because only errors that are captured with the `noticeError()` API will be scoped. Others will will be reported to the container browser application entity.
